xarray/core/dataset.py

```diff
@@ -102,6 +103,105 @@ def calculate_dimensions(variables):
     return dims
 
 
+def merge_indexes(indexes, variables, coord_names, append=False):
```

Comment:
Can you try adding type annotations here, like the ones I started adding in `core/merge.py`? I'm not even running mypy yet but I think these could significantly improve readability, and are lighter weight than adding a full docstring.